map/chart_generator.cpp

```diff
@@ -0,0 +1,215 @@
+#include "map/chart_generator.hpp"
+
+#include "base/assert.hpp"
+
+#include "std/algorithm.hpp"
+#include "std/fstream.hpp"
+
+#include "3party/agg/agg_conv_curve.h"
+#include "3party/agg/agg_conv_stroke.h"
+#include "3party/agg/agg_path_storage.h"
+#include "3party/agg/agg_pixfmt_rgba.h"
+#include "3party/agg/agg_rasterizer_scanline_aa.h"
+#include "3party/agg/agg_renderer_primitives.h"
+#include "3party/agg/agg_renderer_scanline.h"
+#include "3party/agg/agg_scanline_p.h"
+
+namespace
+{
+template <class Color, class Order>
+struct BlendAdaptor
+{
+  using order_type = Order;
+  using color_type = Color;
+  using TValueType = typename color_type::value_type;
+  using TCalcType = typename color_type::calc_type;
+
+  enum
+  {
+    ScaleShift = color_type::base_shift,
+    ScaleMask = color_type::base_mask,
+  };
+
+  static AGG_INLINE void blend_pix(unsigned op, TValueType * p, unsigned cr, unsigned cg,
+                                   unsigned cb, unsigned ca, unsigned cover)
+  {
+    using TBlendTable = agg::comp_op_table_rgba<Color, Order>;
+    if (p[Order::A])
+    {
+      TBlendTable::g_comp_op_func[op](p, (cr * ca + ScaleMask) >> ScaleShift,
+                                      (cg * ca + ScaleMask) >> ScaleShift,
+                                      (cb * ca + ScaleMask) >> ScaleShift, ca, cover);
+    }
+    else
+    {
+      TBlendTable::g_comp_op_func[op](p, cr, cg, cb, ca, cover);
+    }
+  }
+};
+}  // namespace
+
+void NormalizeChartData(deque<double> const & distanceDataM,
+                        feature::TAltitudes const & altitudeDataM, size_t resultPointCount,
+                        vector<double> & uniformAltitudeDataM)
+{
+  uniformAltitudeDataM.clear();
+  if (distanceDataM.empty() || resultPointCount == 0 ||
+      distanceDataM.size() != altitudeDataM.size())
+    return;
+
+  if (!is_sorted(distanceDataM.cbegin(), distanceDataM.cend()))
+  {
+    LOG(LERROR, ("Route segment distances are not sorted."));
+    return;
+  }
+
+  uniformAltitudeDataM.resize(resultPointCount);
+  double const routeLenM = distanceDataM.back();
+  double const stepLenM = routeLenM / static_cast<double>(resultPointCount - 1);
+
+  auto const calculateAltitude = [&](double distFormStartM) {
+    if (distFormStartM <= distanceDataM.front())
+      return static_cast<double>(altitudeDataM.front());
+    if (distFormStartM >= distanceDataM.back())
+      return static_cast<double>(altitudeDataM.back());
+
+    auto const lowerIt = lower_bound(distanceDataM.cbegin(), distanceDataM.cend(), distFormStartM);
```

Comment:
optional. you can speed up algorithm: since distanceFromStartM is monotonic growing, it should be enough to lower_bound(prevLowerId, distanceDataM.end()), where prevLowerId initially set to distanceDataM.begin().